LiveSPICEVst/LiveSPICEPlugin.cs

```diff
@@ -0,0 +1,322 @@
+            {
+                Logging.Log("Load program failed: " + ex.Message);
+            }
+        }
+
+        public void LoadSchematic(string path)
+        {
+            haveSimulationError = false;
+
+            try
+            {
+                SimulationProcessor.LoadSchematic(path);
+            }
+            catch (Exception ex)
+            {
+                MessageBox.Show(String.Format("Error loading schematic from: {0}\n\n{1}", path, ex.Message), "Schematic Load Error", MessageBoxButton.OK, MessageBoxImage.Error);
+            }
+        }
+
+        /// <summary>
+        /// Process a buffer of samples from the VST host
+        /// </summary>
+        /// <param name="input">IntPtr to a double** of input samples</param>
+        /// <param name="output">IntPtr to a double** of output samples</param>
+        /// <param name="inChannelCount">The number of input channels</param>
+        /// <param name="outChannelCount">The number of output channels</param>
+        /// <param name="bufferSize">The buffer size in samples</param>
+        public void ProcessSample(IntPtr input, IntPtr output, uint inChannelCount, uint outChannelCount, uint bufferSize)
+        {
+            if (haveSimulationError)
+            {
+                // If we had an error running the simulation, bypass it
+
+                uint bytesToCopy = bufferSize * sizeof(double);
+
+                Buffer.MemoryCopy((void *)((double**)input)[0], (void *)((double**)output)[0], bytesToCopy, bytesToCopy);
+            }
+            else
+            {
+                // Check if our buffer size has changed
+                if (currentBufferSize != bufferSize)
+                {
+                    if (currentBufferSize < bufferSize)
+                    {
+                        currentBufferSize = (int)bufferSize;
+
+                        Array.Resize<double>(ref inputBuffers[0], currentBufferSize);
+                        Array.Resize<double>(ref outputBuffers[0], currentBufferSize);
+                    }
+                    else
+                    {
+                        currentBufferSize = (int)bufferSize;
+                    }
+                }
+
+                // We are doing mono processing, so just grab the left input and output channel
+                IntPtr leftIn = (IntPtr)((double**)input)[0];
+                IntPtr leftOut = (IntPtr)((double**)output)[0];
+
+                Marshal.Copy(leftIn, inputBuffers[0], 0, currentBufferSize);
+
+                try
+                {
+                    SimulationProcessor.RunSimulation(inputBuffers, outputBuffers, currentBufferSize);
+                }
+                catch (Exception ex)
+                {
+                    haveSimulationError = true;
+
+                    new Thread(() =>
+                    {
+                        MessageBox.Show("Error running circuit simulation.\n\n" + ex.Message, "Simulation Error", MessageBoxButton.OK, MessageBoxImage.Error);
```

Comment:
Maybe `EditorView.Dispatcher.InvokeAsync(() => ...);` would work here? That would make the message box appear on the dialog thread (maybe that matters, I'm not sure), and avoid the need for a new thread.

Reply:
The EditorView is not guaranteed to be active (loading a schematic when restoring saved state when the plugin window isn't open in the host, for example). Using a new thread seems to work just fine - if you load the "Wien Bridge Oscillator" schematic, you can see it in operation.

Reply:
Makes sense, the thread is OK with me.